### PR TITLE
[mlir][llvm] Implement ConstantLike for ZeroOp, UndefOp, PoisonOp

### PR DESCRIPTION
**Pull request by guy-david**
These act as constants and should be propagated whenever possible. It is safe to do so for mlir.undef and mlir.poison because they remain "dirty" through out their lifetime and can be duplicated, merged, etc. per the LangRef.

If the direction is good, I'll tend to the broken tests.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td [hunk 1]
```diff
@@ -1037,4 +1037,18 @@ def LLVM_TargetFeaturesAttr : LLVM_Attr<"TargetFeatures", "target_features">
...
+
+def LLVM_UndefAttr : LLVM_Attr<"Undef", "undef"> {
+}
+
+//===----------------------------------------------------------------------===//
+// PoisonAttr
+//===----------------------------------------------------------------------===//
+
+def LLVM_PoisonAttr : LLVM_Attr<"Poison", "poison"> {
+}
```
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:1052]: I think we also want a zero attribute for cases where a zero value cannot be expressed with an integer / float attribute etc.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 2]
```diff
@@ -2555,6 +2555,26 @@ Region *LLVMFuncOp::getCallableRegion() {
   return &getBody();
 }
 
+//===----------------------------------------------------------------------===//
+// UndefOp.
+//===----------------------------------------------------------------------===//
+
+/// Fold a undef operation to a dedicated undef attribute.
+OpFoldResult LLVM::UndefOp::fold(FoldAdaptor) {
+  // TODO: Should fold into ub::*?
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2564]: We want to stay in LLVM dialect world here. So for now I would fold to our own LLVM attributes! Same for Poison.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 3]
```diff
@@ -2568,6 +2588,13 @@ LogicalResult LLVM::ZeroOp::verify() {
   return success();
 }
 
+OpFoldResult LLVM::ZeroOp::fold(FoldAdaptor) {
+  OpFoldResult result = Builder(getContext()).getZeroAttr(getType());
+  if (result)
+    return result;
+  return TypeAttr::get(getType());
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2595]: I think here we want to create a LLVM::ZeroAttr

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 4]
```diff
@@ -3276,6 +3303,11 @@ Operation *LLVMDialect::materializeConstant(OpBuilder &builder, Attribute value,
   if (auto symbol = dyn_cast<FlatSymbolRefAttr>(value))
     if (isa<LLVM::LLVMPointerType>(type))
       return builder.create<LLVM::AddressOfOp>(loc, type, symbol);
+  // If this was folded from an llvm.mlir.zero operation, it should be
+  // materialized as such. However, prefer to use llvm.mlir.constant for
+  // compatible types.
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:3308]: Since we fold integers / floats to integer and float attributes we probably don't need this special handling here and we can just do something like this:

```
if (isa<LLVM::ZeroAttr>(value))
  return builder.create<LLVM::ZeroOp>(loc, type);
if (isa<LLVM::UndefAttr>(value))
  return builder.create<LLVM::UndefOp>(loc, type);
if (isa<LLVM::PoisonAttr>(value))
  return builder.create<LLVM::PoisonOp>(loc, type);
```

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir [hunk 5]
```diff
@@ -101,3 +101,63 @@ llvm.func @addressof_blocks(%arg: i1) -> !llvm.ptr {
 }
 
 llvm.mlir.global constant @foo() : i32
+
+// -----
+
+// CHECK-LABEL: llvm.func @zero
+llvm.func @zero() {
+  // CHECK-NEXT: %[[ZERO:.+]] = llvm.mlir.zero : i32
```
Comment (gysit) [inline mlir/test/Dialect/LLVMIR/constant-folding.mlir:109]: Interesting I though this would be a llvm.constant 0 : i32

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir [hunk 6]
```diff
@@ -101,3 +101,63 @@ llvm.func @addressof_blocks(%arg: i1) -> !llvm.ptr {
...
+  %undef2 = llvm.mlir.undef : i32
+  // CHECK-NEXT: llvm.call @foo(%[[UNDEF]], %[[UNDEF]])
+  llvm.call @foo(%undef1, %undef2) : (i32, i32) -> ()
+  // CHECK-NEXT: llvm.return
+  llvm.return
+}
+
+llvm.func @foo(i32, i32)
+
+// -----
```
Comment (gysit) [inline mlir/test/Dialect/LLVMIR/constant-folding.mlir:150]: There are some more tests that are interesting. I think we can test that a switch that gets two zero attributes folds to just zero (undef / poison would probably have special handling that we do not yet implement in LLVM dialect)... Another interesting test could be if we can fold the zext of an zero constant. Folding an addition would be nice as well but LLVM dialect does not implement such foldings...

**Review by gysit — COMMENTED** [review]
Nice!

I added some comments. I think having this special attributes makes sense but I would add one for Zero as well bcs there are Zero values that cannot be represented with an integer / float attribute. E.g. a struct.

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir (hunk 5 above)
Comment (definelicht) [reply mlir/test/Dialect/LLVMIR/constant-folding.mlir:109]: I've only seen `llvm.mlir.zero` used for `nullptr`, what's the convention here?

Comment (gysit) [reply mlir/test/Dialect/LLVMIR/constant-folding.mlir:109]: It can in principle be used everywhere. I think we want to fold it to integer / float zero attributes if possible but we want to keep it for pointers / structs / arrays where there is no good alternative.

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir (hunk 6 above)
Comment (guy-david) [reply mlir/test/Dialect/LLVMIR/constant-folding.mlir:120]: I've added a test for a `select` of null-pointers. Do you think that testing `zext` is also beneficial? It seems a bit like boiler-plate matrix of tests for each constant-like someone wishes to add.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 7]
```diff
@@ -2568,6 +2586,11 @@ LogicalResult LLVM::ZeroOp::verify() {
   return success();
 }
 
+/// Fold a zero operation to a dedicated zero attribute.
+OpFoldResult LLVM::ZeroOp::fold(FoldAdaptor) {
+  return mlir::LLVM::ZeroAttr::get(getContext());
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2591]: Ah sorry for being unprecise here. The logic that checks if you can build a zero attribute is useful here! If the type of the attribute is an integer for example it is preferable to fold to an integer attribute which then in turn can fold with a follow up ZExt operation.

File: mlir/test/Dialect/GPU/dynamic-shared-memory.mlir [hunk 8]
```diff
@@ -19,83 +19,83 @@ gpu.module @modules {
...
-// CHECK: %[[S22:.+]] = builtin.unrealized_conversion_cast %[[S21]] : !llvm.struct<(ptr<3>, ptr<3>, i64, array<2 x i64>, array<2 x i64>)> to memref<32x64xf32, #gpu.address_space<workgroup>>
-// CHECK: "test.use.shared.memory"(%[[S22]]) : (memref<32x64xf32, #gpu.address_space<workgroup>>) -> ()
+
+// CHECK-DAG: %[[S0:.+]] = llvm.mlir.constant(32 : index) : i64
+// CHECK-DAG: %[[S1:.+]] = llvm.mlir.constant(64 : index) : i64
+// CHECK-DAG: %[[S2:.+]] = llvm.mlir.undef : !llvm.struct<(ptr<3>, ptr<3>, i64, array<2 x i64>, array<2 x i64>)>
+// CHECK-DAG: %[[S3:.+]] = llvm.mlir.constant(1 : index) : i64
+// CHECK-DAG: %[[S4:.+]] = llvm.mlir.constant(0 : index) : i64
+// CHECK-DAG: %[[S5:.+]] = llvm.mlir.addressof @__dynamic_shmem__3 : !llvm.ptr<3>
+//     CHECK: %[[S6:.+]] = llvm.insertvalue %[[S5]], %[[S2]][0] : !llvm.struct<(ptr<3>, ptr<3>, i64, array<2 x i64>, array<2 x i64>)>
```
Comment (gysit) [inline mlir/test/Dialect/GPU/dynamic-shared-memory.mlir:29]:
```suggestion
// CHECK: %[[S6:.+]] = llvm.insertvalue %[[S5]], %[[S2]][0] : !llvm.struct<(ptr<3>, ptr<3>, i64, array<2 x i64>, array<2 x i64>)>
```
nit: did you see this indention somewhere else? Usually I would not put extra whitespaces except there is another test in the file that does so.

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir (hunk 6 above)
Comment (gysit) [reply mlir/test/Dialect/LLVMIR/constant-folding.mlir:120]: If we fold mlir.zero : i32 to an integer attribute as mentioned above then there would be some value since the zero operation should disappear and fold with the zext.

File: mlir/test/Dialect/SparseTensor/specifier_to_llvm.mlir [hunk 9]
```diff
@@ -3,12 +3,12 @@
...
 
 // CHECK-LABEL:   func.func @sparse_metadata_init() -> !llvm.struct<(array<2 x i64>, array<3 x i64>)> {
-// CHECK:           %[[VAL_0:.*]] = arith.constant 0 : i64
-// CHECK:           %[[VAL_1:.*]] = llvm.mlir.undef : !llvm.struct<(array<2 x i64>, array<3 x i64>)>
-// CHECK:           %[[VAL_2:.*]] = llvm.insertvalue %[[VAL_0]], %[[VAL_1]][1, 0] : !llvm.struct<(array<2 x i64>, array<3 x i64>)>
-// CHECK:           %[[VAL_3:.*]] = llvm.insertvalue %[[VAL_0]], %[[VAL_2]][1, 1] : !llvm.struct<(array<2 x i64>, array<3 x i64>)>
-// CHECK:           %[[VAL_4:.*]] = llvm.insertvalue %[[VAL_0]], %[[VAL_3]][1, 2] : !llvm.struct<(array<2 x i64>, array<3 x i64>)>
-// CHECK:           return %[[VAL_4]] : !llvm.struct<(array<2 x i64>, array<3 x i64>)>
+// CHECK-DAG:       %[[STRUCT:.*]] = llvm.mlir.undef : !llvm.struct<(array<2 x i64>, array<3 x i64>)>
+// CHECK-DAG:       %[[ZERO:.*]] = arith.constant 0 : i64
```
Comment (gysit) [inline mlir/test/Dialect/SparseTensor/specifier_to_llvm.mlir:7]:
```suggestion
// CHECK-DAG:       %[[CST0:.*]] = arith.constant 0 : i64
```
nit: there is a big diversity in how to name this check variables but CST0 would be closer to what is usually used than ZERO.

**Review by gysit — COMMENTED** [review]
Cool I think I was confusing you with the first comment :). The conversion to an integer attribute in the folder of the ZeroOp is fine. However, I would just use this when possible and otherwise we need a ZeroAttr. Feel free to reach out if my comments are not self explanatory.

File: mlir/test/Dialect/GPU/dynamic-shared-memory.mlir (hunk 8 above)
Comment (guy-david) [reply mlir/test/Dialect/GPU/dynamic-shared-memory.mlir:29]: Yes, ctrl+f `   CHECK:` and you'll find many like this.

Comment (gysit) [reply mlir/test/Dialect/GPU/dynamic-shared-memory.mlir:29]: Ok then let's keep it...

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 10]
```diff
@@ -3271,11 +3298,17 @@ LogicalResult LLVMDialect::verifyRegionResultAttribute(Operation *op,
 
 Operation *LLVMDialect::materializeConstant(OpBuilder &builder, Attribute value,
                                             Type type, Location loc) {
-  // If this was folded from an llvm.mlir.addressof operation, it should be
-  // materialized as such.
+  // If this was folded from a ConstantLike operation, it should be materialized
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:3301]:
```suggestion
  // If this was folded from another operation than llvm.mlir.constant op, it should be materialized
```
nit: the constant op is also ConstantLike. I think I would explicitly say not mlir.constant somehow.

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir [hunk 11]
```diff
@@ -101,3 +101,75 @@ llvm.func @addressof_blocks(%arg: i1) -> !llvm.ptr {
...
+  // CHECK-NEXT: %[[ZERO:.+]] = llvm.mlir.constant(0 : i64) : i64
+  %zero = llvm.mlir.zero : i32
+  %zero_extended = llvm.zext %zero : i32 to i64
+  // CHECK-NEXT: llvm.return %[[ZERO]]
+  llvm.return %zero_extended : i64
+}
+
+// -----
+
+llvm.func @foo(!llvm.ptr, !llvm.ptr)
```
Comment (gysit) [inline mlir/test/Dialect/LLVMIR/constant-folding.mlir:165]:
```suggestion
```

File: mlir/test/Dialect/LLVMIR/constant-folding.mlir [hunk 12]
```diff
@@ -101,3 +101,75 @@ llvm.func @addressof_blocks(%arg: i1) -> !llvm.ptr {
...
+  %nullptr2 = llvm.mlir.zero : !llvm.ptr
+  // CHECK-NEXT: llvm.call @foo(%[[NULLPTR]], %[[NULLPTR]])
+  llvm.call @foo(%nullptr1, %nullptr2) : (!llvm.ptr, !llvm.ptr) -> ()
+  // CHECK-NEXT: llvm.return
+  llvm.return
+}
+
+// -----
+
+llvm.func @foo(!llvm.ptr, !llvm.ptr)
```
Comment (gysit) [inline mlir/test/Dialect/LLVMIR/constant-folding.mlir:152]:
```suggestion
```

**Review by gysit — COMMENTED** [review]
LGTM modulo some small nits.

Let's prepare an upstream PR?

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td [hunk 13]
```diff
@@ -1037,4 +1037,25 @@ def LLVM_TargetFeaturesAttr : LLVM_Attr<"TargetFeatures", "target_features">
   let genVerifyDecl = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// UndefAttr
+//===----------------------------------------------------------------------===//
+
+def LLVM_UndefAttr : LLVM_Attr<"Undef", "undef"> {
```
Comment (definelicht) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:1044]: Maybe comment on why these exists -- to make them foldable?

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td [hunk 14]
```diff
@@ -1037,4 +1037,25 @@ def LLVM_TargetFeaturesAttr : LLVM_Attr<"TargetFeatures", "target_features">
...
+}
+
+//===----------------------------------------------------------------------===//
+// PoisonAttr
+//===----------------------------------------------------------------------===//
+
+def LLVM_PoisonAttr : LLVM_Attr<"Poison", "poison"> {
```
Comment (definelicht) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:1051]: Maybe comment on why these exists -- to make them foldable?

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td [hunk 15]
```diff
@@ -1037,4 +1037,25 @@ def LLVM_TargetFeaturesAttr : LLVM_Attr<"TargetFeatures", "target_features">
...
+}
+
+//===----------------------------------------------------------------------===//
+// ZeroAttr
+//===----------------------------------------------------------------------===//
+
+def LLVM_ZeroAttr : LLVM_Attr<"Zero", "zero"> {
```
Comment (definelicht) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:1058]: Maybe comment on why these exists -- to make them foldable?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 16]
```diff
@@ -2568,6 +2586,15 @@ LogicalResult LLVM::ZeroOp::verify() {
   return success();
 }
 
+/// Fold a zero operation to a builtin zero attribute when possible and fallback
```
Comment (definelicht) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:2589]:
```suggestion
/// Fold a zero operation to a builtin zero attribute when possible and fall back
```

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp (hunk 10 above)
Comment (guy-david) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:3301]: I've also added a comment on ZeroOp specifically which may be folded into the built-in zero attribute.
